ledger/simulation/trace.go

```diff
@@ -67,16 +69,76 @@ func makeTxnGroupResult(txgroup []transactions.SignedTxn) TxnGroupResult {
+type ResultEvalConstantsBuilder struct {
+	Result ResultEvalConstants
+}
+
+// NewResultEvalConstantsBuilder constructs a builder for ResultEvalConstants
+func NewResultEvalConstantsBuilder() *ResultEvalConstantsBuilder {
+	return &ResultEvalConstantsBuilder{}
+}
+
+// SimulateLogBytesLimit hardcode limit of how much bytes one can log during simulation (with lift-log-limits)
+const SimulateLogBytesLimit = uint64(65536)
+
+// LiftLogLimits method modify the log limits from lift option:
+// - if lift log limits, then overload result from local Config
+// - otherwise, set `LogLimits` field to be nil
+func (r *ResultEvalConstantsBuilder) LiftLogLimits(lift bool) *ResultEvalConstantsBuilder {
+	if lift {
+		maxLogCalls, maxLogSize := uint64(config.MaxLogCalls), SimulateLogBytesLimit
+		r.Result = ResultEvalConstants{
+			MaxLogCalls: &maxLogCalls,
+			MaxLogSize:  &maxLogSize,
+		}
+	}
+	return r
+}
+
+// Finalize method cleanup the *ResultEvalConstants if it is empty ResultEvalConstants{}, then return nil,
+// otherwise it returns the pointer to ResultEvalConstants
+func (r *ResultEvalConstantsBuilder) Finalize() *ResultEvalConstants {
+	// Since ResultEvalConstants is omitempty, we want to check if it is actually empty
+	if r.Result == (ResultEvalConstants{}) {
+		return nil
+	}
+	return &r.Result
+}
+
+// LogicEvalConstants method infers the logic.EvalConstants from Result.EvalConstants (*ResultEvalConstants)
+// and generate appropriate parameters to override during simulation runtime.
+func (c *ResultEvalConstants) LogicEvalConstants() logic.EvalConstants {
+	logicEvalConstants := logic.RuntimeEvalConstants()
+	if c == nil {
+		return logicEvalConstants
+	}
+	if c.MaxLogSize != nil {
+		logicEvalConstants.MaxLogSize = *c.MaxLogSize
+	}
+	if c.MaxLogCalls != nil {
+		logicEvalConstants.MaxLogCalls = *c.MaxLogCalls
+	}
+	return logicEvalConstants
+}
+
 // Result contains the result from a call to Simulator.Simulate
 type Result struct {
-	Version      uint64
-	LastRound    basics.Round
-	TxnGroups    []TxnGroupResult // this is a list so that supporting multiple in the future is not breaking
-	WouldSucceed bool             // true iff no failure message, no missing signatures, and the budget was not exceeded
-	Block        *ledgercore.ValidatedBlock
+	Version       uint64
+	LastRound     basics.Round
+	TxnGroups     []TxnGroupResult // this is a list so that supporting multiple in the future is not breaking
+	WouldSucceed  bool             // true iff no failure message, no missing signatures, and the budget was not exceeded
+	EvalConstants *ResultEvalConstants
+	Block         *ledgercore.ValidatedBlock
 }
 
-func makeSimulationResultWithVersion(lastRound basics.Round, txgroups [][]transactions.SignedTxn, version uint64) (Result, error) {
+func makeSimulationResultWithVersion(lastRound basics.Round, txgroups [][]transactions.SignedTxn, version uint64, liftLogLimits bool) (Result, error) {
```

Comment:
Let's discuss a bit under this thread: I was thinking of making `Request` take `txn-groups`, rather than current `txn-group`. This will be helpful for later implementation, what do you think?

Reply:
That sounds good, as long as there's a check to ensure there's only a single group

Reply:
see [take Request in simulator internal functions](https://github.com/algorand/go-algorand/pull/5247/commits/b5d64405c7660113ea0367fa25b3e3835e05ad8f)